### DISCUSSION TITLE
Eliminate the requirement to define attributes!

### DISCUSSION THREAD

**Discussion by natgross** [appwrite/appwrite#3616] [General]
Come on guys. Is this a true document db with collections, or are you simply wrapping SQL tables with document lingo? I have to define fields and types per collection (ahem- table)??? I forgot how to do that when I dropped Oracle DB in favor of document db's 10 years ago! (Oh... The good old days, sql/j2ee/jpa)
Is there any plan for a true document based db on appwrite?
Thanks,
nat.

**Comment by BenUniqcode** [comment]
I have to say, this is one of my biggest bugbears about Appwrite at the moment: It's the worst of both worlds!
It has the fixed, restrictive schema of an RDBMS but with none of the advantages of proper relations, joins, or optimised queries.
Pick one: Either be NoSQL and schemaless, or be a schema-driven RDBMS but actually store the data properly and make use of the decades of RDBMS optimisations instead of reinventing them (like Directus for example). Which way are you going to go?

**Reply by natgross** [reply]
When I was reviewing the docs (before actually installing Appwrite) and skimmed the pages on attributes, I was thinking 'huh?', but quickly (and incorrectly) assumed that aw gives you an option of defining required attributes. Nice.
But, when I tried it, woa, I couldn't believe the exception. 
It's not only semantics. Adding/dropping a field with a 'collection' [table] of a million plus 'documents' [rows], and indices, gimme a break.

This is not meant as a discussion of sql versus noSql, they both have their place in society. My gripe [at the moment] is why hide a sql database behind a document based wrapper.  Why waste so much of my time? (albeit, at the end, I'm glad I installed aw due to the other goodies it offers.)

**Reply by stnguyen90** [reply]
Regarding the lack of joins, we do understand this is a pain point and are working towards making that better. GraphQL could help with this or maybe more advanced queries in the future. Our biggest priority, however, is to uphold the highest level of performance we can.

**Comment by stnguyen90** [comment]
Appwrite aims to be an abstraction on top of underlying infrastructure. That's why we're working towards adding support for using any database (MongoDB, Postgres, Cockroach, etc.). Given that, I don't think we'll be planning on supporting unstructured documents.

Although we won't change to unstructured documents, we may introduce NoSQL features once MongoDB support is released and developers choose to use a MongoDB database.

**Reply by natgross** [reply]
It's good to hear that the plan is for more databases. This is a HUGE undertaking. All the best.
